File: frontend/src/api/ingestion.ts
import type {
  AIReportDraftResponse,
  AIReportDraftSaveResponse,
  AIQueryResponse,
  BatchSummary,
  IngestionRunResponse,
  ValidationErrorCount,
  ValidationErrorDetail,
  ValidationErrorRow,
} from "../types/ingestion";

async function parseJsonResponse<T>(response: Response): Promise<T> {
  if (!response.ok) {
    let detail = response.statusText;
    try {
      const body = (await response.json()) as { detail?: string };
      detail = body.detail ?? detail;
    } catch {
      // Ignore JSON parsing issues and fall back to the status text.
    }
    throw new Error(detail || "Request failed.");
  }

  return (await response.json()) as T;
}

export async function uploadWorkbook(file: File): Promise<IngestionRunResponse> {
  return uploadWorkbookWithProgress(file);
}

export async function uploadWorkbookWithProgress(
  file: File,
  onProgress?: (percent: number) => void,
): Promise<IngestionRunResponse> {
  const formData = new FormData();
  formData.append("file", file);

  return new Promise<IngestionRunResponse>((resolve, reject) => {
    const request = new XMLHttpRequest();
    request.open("POST", "/api/ingestion/upload");

    request.upload.onprogress = (event) => {
      if (!onProgress || !event.lengthComputable) {
        return;
      }
      const percent = Math.min(100, Math.round((event.loaded / event.total) * 100));
      onProgress(percent);
    };

    request.onerror = () => {
      reject(new Error("Upload failed."));
    };

    request.onload = () => {
      try {
        const responseText = request.responseText || "{}";
        const parsed = JSON.parse(responseText) as IngestionRunResponse & { detail?: string };

        if (request.status >= 200 && request.status < 300) {
          onProgress?.(100);
          resolve(parsed);
          return;
        }

        reject(new Error(parsed.detail ?? request.statusText ?? "Upload failed."));
      } catch {
        reject(new Error(request.statusText || "Upload failed."));
      }
    };

    request.send(formData);
  });
}

export async function getBatchSummary(loadBatchId: string): Promise<BatchSummary> {
  const response = await fetch(`/api/batches/${loadBatchId}/summary`);
  return parseJsonResponse<BatchSummary>(response);
}

export async function getBatchErrorCounts(
  loadBatchId: string,
): Promise<ValidationErrorCount[]> {
  const response = await fetch(`/api/batches/${loadBatchId}/error-counts`);
  return parseJsonResponse<ValidationErrorCount[]>(response);
}

export async function getBatchErrorDetails(
  loadBatchId: string,
): Promise<ValidationErrorDetail[]> {
  const response = await fetch(`/api/batches/${loadBatchId}/error-details`);
  return parseJsonResponse<ValidationErrorDetail[]>(response);
}

export async function getBatchErrorRows(
  loadBatchId: string,
): Promise<ValidationErrorRow[]> {
  const response = await fetch(`/api/batches/${loadBatchId}/error-rows`);
  return parseJsonResponse<ValidationErrorRow[]>(response);
}

export function getErrorDownloadUrl(loadBatchId: string): string {
  return `/api/batches/${loadBatchId}/download-errors`;
}

export async function runAIQuery(question: string): Promise<AIQueryResponse> {
  const response = await fetch("/api/ai/query", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ question }),
  });
  return parseJsonResponse<AIQueryResponse>(response);
}

export async function createAIReportDraft(
  projectId: string,
  options?: { regenerateFresh?: boolean },
): Promise<AIReportDraftResponse> {
  const response = await fetch("/api/ai/report-draft", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      project_id: projectId,
      regenerate_fresh: Boolean(options?.regenerateFresh),
    }),
  });
  return parseJsonResponse<AIReportDraftResponse>(response);
}

export async function saveAIReportDraft(payload: {
  project_id?: string | null;
  load_batch_id: string;
  source_file_name?: string | null;
  draft_sections: Record<string, unknown>;
}): Promise<AIReportDraftSaveResponse> {
  const response = await fetch("/api/ai/report-draft/save", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  return parseJsonResponse<AIReportDraftSaveResponse>(response);
}

async function exportAIReport(
  endpoint: string,
  payload: {
    project_id?: string | null;
    load_batch_id: string;
    source_file_name?: string | null;
    draft_sections: Record<string, unknown>;
    report_context: Record<string, unknown>;
  },
): Promise<void> {
  const response = await fetch(endpoint, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  if (!response.ok) {
    let detail = response.statusText;
    try {
      const body = (await response.json()) as { detail?: string };
      detail = body.detail ?? detail;
    } catch {
      // no-op
    }
    throw new Error(detail || "Export failed.");
  }
  const blob = await response.blob();
  const disposition = response.headers.get("Content-Disposition") || "";
  const filenameMatch = disposition.match(/filename="?([^"]+)"?/i);
  const fileName = filenameMatch?.[1] ?? "report";
  const url = URL.createObjectURL(blob);
  const a = document.createElement("a");
  a.href = url;
  a.download = fileName;
  document.body.appendChild(a);
  a.click();
  a.remove();
  URL.revokeObjectURL(url);
}

export async function exportAIReportDocx(payload: {
  project_id?: string | null;
  load_batch_id: string;
  source_file_name?: string | null;
  draft_sections: Record<string, unknown>;
  report_context: Record<string, unknown>;
}): Promise<void> {
  await exportAIReport("/api/ai/report-export/docx", payload);
}

export async function exportAIReportPdf(payload: {
  project_id?: string | null;
  load_batch_id: string;
  source_file_name?: string | null;
  draft_sections: Record<string, unknown>;
  report_context: Record<string, unknown>;
}): Promise<void> {
  await exportAIReport("/api/ai/report-export/pdf", payload);
}

File: backend/app/api/routes/ingestion.py
from fastapi import APIRouter, File, UploadFile
from fastapi.responses import FileResponse, StreamingResponse

from ...schemas.ingestion import (
    AIReportExportRequest,
    AIReportDraftSaveRequest,
    AIReportDraftSaveResponse,
    AIReportDraftRequest,
    AIReportDraftResponse,
    AIQueryRequest,
    AIQueryResponse,
    BatchSummaryResponse,
    IngestionRunResponse,
    ValidationErrorCountResponse,
    ValidationErrorDetailResponse,
    ValidationErrorRowResponse,
)
from ...services.ai_query_service import run_ai_query
from ...services.ai_report_service import build_report_draft, save_report_draft_state
from ...services.report_export_service import export_report_docx, export_report_pdf
from ...services.ingestion_service import (
    build_batch_error_csv,
    get_batch_error_counts,
    get_batch_error_details,
    get_batch_error_rows,
    get_batch_summary,
    run_ingestion_from_upload,
)

router = APIRouter(tags=["ingestion"])


@router.post("/ingestion/upload", response_model=IngestionRunResponse)
async def trigger_ingestion_from_upload(
    file: UploadFile = File(...),
) -> IngestionRunResponse:
    result = await run_ingestion_from_upload(file)
    return IngestionRunResponse(**result)


@router.get("/batches/{load_batch_id}/summary", response_model=BatchSummaryResponse)
def fetch_batch_summary(load_batch_id: str) -> BatchSummaryResponse:
    summary = get_batch_summary(load_batch_id)
    return BatchSummaryResponse(**summary)


@router.get(
    "/batches/{load_batch_id}/error-counts",
    response_model=list[ValidationErrorCountResponse],
)
def fetch_batch_error_counts(load_batch_id: str) -> list[ValidationErrorCountResponse]:
    counts = get_batch_error_counts(load_batch_id)
    return [ValidationErrorCountResponse(**row) for row in counts]


@router.get(
    "/batches/{load_batch_id}/error-details",
    response_model=list[ValidationErrorDetailResponse],
)
def fetch_batch_error_details(load_batch_id: str) -> list[ValidationErrorDetailResponse]:
    details = get_batch_error_details(load_batch_id)
    return [ValidationErrorDetailResponse(**row) for row in details]


@router.get(
    "/batches/{load_batch_id}/error-rows",
    response_model=list[ValidationErrorRowResponse],
)
def fetch_batch_error_rows(load_batch_id: str) -> list[ValidationErrorRowResponse]:
    rows = get_batch_error_rows(load_batch_id)
    return [ValidationErrorRowResponse(**row) for row in rows]


@router.get("/batches/{load_batch_id}/download-errors")
def download_batch_error_details(load_batch_id: str) -> StreamingResponse:
    csv_buffer = build_batch_error_csv(load_batch_id)
    filename = f"batch-{load_batch_id}-errors.csv"
    headers = {"Content-Disposition": f'attachment; filename="{filename}"'}
    return StreamingResponse(csv_buffer, media_type="text/csv", headers=headers)


@router.post("/ai/query", response_model=AIQueryResponse)
def run_ai_sql_query(payload: AIQueryRequest) -> AIQueryResponse:
    result = run_ai_query(payload.question)
    return AIQueryResponse(**result)


@router.post("/ai/report-draft", response_model=AIReportDraftResponse)
def create_ai_report_draft(payload: AIReportDraftRequest) -> AIReportDraftResponse:
    result = build_report_draft(
        project_id=payload.project_id,
        load_batch_id=payload.load_batch_id,
        use_saved_draft=not payload.regenerate_fresh,
    )
    return AIReportDraftResponse(**result)


@router.post("/ai/report-draft/save", response_model=AIReportDraftSaveResponse)
def save_ai_report_draft(payload: AIReportDraftSaveRequest) -> AIReportDraftSaveResponse:
    result = save_report_draft_state(
        load_batch_id=payload.load_batch_id,
        project_id=payload.project_id,
        source_file_name=payload.source_file_name,
        draft_sections=payload.draft_sections,
    )
    return AIReportDraftSaveResponse(**result)


@router.post("/ai/report-export/docx")
def export_ai_report_docx(payload: AIReportExportRequest) -> FileResponse:
    output_path = export_report_docx(payload.model_dump())
    return FileResponse(
        path=output_path,
        media_type="application/vnd.openxmlformats-officedocument.wordprocessingml.document",
        filename=output_path.name,
    )


@router.post("/ai/report-export/pdf")
def export_ai_report_pdf(payload: AIReportExportRequest) -> FileResponse:
    output_path = export_report_pdf(payload.model_dump())
    return FileResponse(
        path=output_path,
        media_type="application/pdf",
        filename=output_path.name,
    )
